### ISSUE TITLE
Cannot do unit testing with the way we reference our code

### ISSUE THREAD

**Issue by cmbrose** [lighthouse-isu/lighthouse#16]
**Issue:**
With the way our packages reference each other, we are severely limited in our capabilities for unit tests.

**Cause:**
Our packages import directly from the Go's github.com folder.  Any local changes we make are not reflected in that code.  As a result, we cannot fix bugs while testing (without fully merging our potentially broken code...).

**Potential Solution:** 
We could reference packages by relative paths again.  This will allow our tests to reference our local code and keep them completely separate from the github imports.

**Notes:**
For some tests, using relative paths only in the tests is enough to work with local changes.  However, if the referenced code in turn references something else, then we will have the same issue.

**Example:**

Our package structure currently is:

```
lighthouse/
    lighthouse.go
    handlers/
        handlers.go
        handlers_test.go
```

In handlers/handlers_test.go:

``` go
package handlers
import ("..") // main package

func Test_1(...) {
    //...
    main.Method1();
}

func Test_2(...) {
    //...
    main.Method2();
}
```

In lighthouse.go:

``` go
package main
import ("github.com/lighthouse/lighthouse/handlers")

func Method1() {
    // This code is modified
}

func Method2() {
    //...
    handlers.Method3();
}
```

In handlers/handlers.go:

``` go
package handlers

func Method3() {
    // This code is modified
}
```

Here, our tests will reflect changes made in `lighthouse.go`, but will not reflect changes made in `handlers.go` (which are referenced through `lighthouse.go`).

I know there is a lot of redirection through the packages here, but mostly it is setup for tests.  For example, I might call into the `main` package to setup the mux routers to test a function that extracts url variables.

**Comment by cmbrose** [comment]
@theMagicalKarp this package structure was mostly your work.  Do you see a clean way of fixing this?

@chfogerty @ngmiller @zach-taylor any input from you guys is appreciated as well.

**Comment by robertsheehy-wf** [comment]
@cmbrose which directory are you making code changes to?  With the "great package migration" any new changes your making to lighthouse should be in `$GOPATH/src/lighthouse/lighthouse`.  That directory is a working local git repository. Also any changes you make in that directory is reflected locally when you import anything from `github.com/lighthouse/lighthouse`.  

Also all the changes inside `$GOPATH/src/lighthouse/lighthouse` can be built with `go get github.com/lighthouse/lighthouse/`.

**Comment by cmbrose** [comment]
@theMagicalKarp So just to clarify, we can make changes directly in the `$GOPATH/src/github.com/lighthouse/lighthouse` directory, and we don't have to worry about it being blown up if you do another `go get`?

If that's the case then this issue is not actually an issue.

**Comment by theMagicalKarp** [comment]
@cmbrose Yup, `go get` retrieves the package from github if it doesn't already exist locally in the gopath and then it will compile the package into a executable.  If the package does exists locally, it will use all the local stuff, including any changes that might have been made to it.  Alternatively you might be able to use `go install github.com/lighthouse/lighthouse` to bypass that github check, but I'm not for sure. All of this information is detailed [here](https://golang.org/cmd/go/).  That also covers how the `go test` command works. I also mocked up a how to setup/run [here](https://gist.github.com/theMagicalKarp/749dc873331f81022f67), I just learned about most of this stuff last week.
